pyvo/dal/adhoc.py

```diff
@@ -154,12 +168,68 @@ class DatalinkResultsMixin(AdhocServiceResultsMixin):
     """
     Mixing for datalink functionallity for results classes.
     """
+
     def iter_datalinks(self):
         """
         Iterates over all datalinks in a DALResult.
         """
-        for record in self:
-            yield record.getdatalink()
+        # To reduce the number of calls to the Datalink service, multiple
+        # IDs are sent in batches. The appropriate batch size is not available
+        # as each service has its own criteria for determining the maximum
+        # number of IDs processed per request. To overcome this limitation,
+        # this method initially sends all the available IDs and if the
+        # response is partial, uses the size of the response as the batch
+        # size.
+        if not hasattr(self, '_datalink'):
+            try:
+                self._datalink = self.get_adhocservice_by_ivoid(DATALINK_IVOID)
+            except DALServiceError:
+                self._datalink = None
+        if not hasattr(self, "_current_batch"):
+            # static vars
+            self._remaining_ids = []  # remaining IDs to processed
+            self._current_batch = None  # retrieved but not returned yet
+            self._current_ids = []  # retrieved but not returned
+            self._processed_ids = []  # retrived and returned IDs
+            self._batch_size = None  # size of the batch
+
+        for row in self:
+            if self._datalink:
+                if not self._current_ids:
+                    if self._batch_size is None:
+                        # first call.
+                        self.query = DatalinkQuery.from_resource(
+                            [_ for _ in self], self._datalink)
+                        self._remaining_ids = self.query['ID']
+                    if not self._remaining_ids:
+                        # we are done
+                        return
+                    if self._batch_size:
+                        # subsequent calls are limitted to batch size
+                        self.query['ID'] = \
+                            self._remaining_ids[:self._batch_size]
+                    self._current_batch = self.query.execute(post=True)
+                    self._current_ids = list(OrderedDict.fromkeys(
+                        [_ for _ in self._current_batch.to_table()['ID']]))
+                    if not self._current_ids:
+                        raise DALService(
+                            'Could not retrieve datalinks for: {}'.format(
+                                ', '.join([_ for _ in
+                                           self._remaining_ids])))
+                    self._batch_size = len(self._current_ids)
+                id = self._current_ids.pop(0)
+                self._processed_ids.append(id)
+                self._remaining_ids.remove(id)
+                yield self._current_batch.clone_byid(id)
+            else:
+                yield DatalinkResults.from_result_url(row.getdataurl())
```

Comment:
Umm. No. Please stick to `getdatalink` because of parameters which are described in the `RESOURCE` element.

Reply:
That was wrong. I've replaced it with: `row.access_format == 'application/x-votable+xml;content=datalink'` (according to the specs)

Reply:
Hmm. I think there should be an `else` clause, even if it just emits warnings.

Reply:
`else: yield None` maybe since there is no info to determine the corresponding datalink for that row?

Reply:
tbh I'm not sure whats the best way here. There shouldn't be empty rows but they also shouldn't go unnoticed.

Reply:
The iterator is returning Datalink resources associated with each row. Question is what to do when no such resources are available (access_url is probably a direct url in that case). This case is valid so I don't think the method should err or warn. The question IMO is whether to return `None` or continue and now that I'm thinking more about it I'm more inclined for the latter (`None` in itself is not a useful info). The contract of the method is to return corresponding Datalink resources so skipping the rows that don't have such resources is OK, no?

This is the relevant section of SIA spec:
```3.1.3 Use of access_url and access_format
If the SIA service is only dealing with simple data (one file per result), the 
access_url column may be a link directly to that file, in which case the 
access_format column should specify the file format (e.g. application/fits).

If the data provider implements a DataLink service for the data being found 
via the SIA {query} capability, they may put a URL to invoke the DataLink 
{links} capability (with ID parameter and value) in the access_url column; if 
they do this, they must also put the standard DataLink MIME type [9] in the 
access_format column.```

Reply:
So, better to return `None` instead of `continue` you think?